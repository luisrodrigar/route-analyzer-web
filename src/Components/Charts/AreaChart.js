import React from 'react';
import AbstractChart from './AbstractChart';
import DateArea from './DataArea';

export class AreaChart extends React.Component{

	render(){
		return( 
			<AbstractChart {...this.props}  >
				{ this.props.data && this.props.data.length && 
	              	<DateArea   {...this.props}/>
				}
			</AbstractChart>
		);
	}
}

export default AreaChart;