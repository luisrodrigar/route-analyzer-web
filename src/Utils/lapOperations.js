import randomColor from 'randomcolor'; 

  // Unused because It delegates calculating the zoom to fitBound in RouteMapComponent (zoomToMarkers)  
  export function getCenterLaps(laps){
    let resumeCenterByLap = laps.map(lap=>{
        return getCenterOfALap(lap);
      });
    const lat = resumeCenterByLap.reduce((total, center)=>{
      return total = total + parseFloat(center.lat);
    }, 0.0)/laps.length;
    const lng = resumeCenterByLap.reduce((total, center)=>{
      return total = total + parseFloat(center.lng);
    }, 0.0)/laps.length;

    return {
      lat,
      lng
    }
  }

  export function getLapsTrackPoints(laps){
    return laps.map(lap => {
        return {
              index:lap.index,
              tracks:getLapTrackPoint(lap),
              color:randomColor(),
              startTime: lap.startTime,
              totalTime: lap.totalTimeSeconds,
              distance: lap.distanceMeters,
              maxSpeed: lap.maximunSpeed,
              avgSpeed: lap.averageSpeed,
              avgBpm: lap.averageHearRate,
              maxBpm: lap.maximunHeartRate,
              cal: lap.calories,
              intensity: lap.intensity
        }
    });
  }

  function getLapTracksWithPosition(lap){
    return lap.tracks
      .filter(track => {
        if(track && track.position)
          return track;
        else return null;
      });
  }

  function getLapTrackPoint(lap){
    return getLapTracksWithPosition(lap).map(trackPoint => {
        var lat = parseFloat(trackPoint.position.latitudeDegrees);
        var lng = parseFloat(trackPoint.position.longitudeDegrees);
        var alt = parseFloat(trackPoint.altitudeMeters);
        var speed = parseFloat(trackPoint.speed);
        var date = parseInt(trackPoint.date,10);
        var dist = parseFloat(trackPoint.distanceMeters);
        var bpm = parseInt(trackPoint.heartRateBpm,10);
        var index = parseInt(trackPoint.index,10);
        return {
          index,
          date,
          position:{
            lat,
            lng
          },
          alt,
          speed,
          dist,
          bpm
        };
    });
  }


  function getCenterOfALap(lap){
    const tracksPositionNotNull = getLapTracksWithPosition(lap);
    const lat = tracksPositionNotNull
      .map(track =>{
        return parseFloat(track.position.latitudeDegrees);
      }).reduce(
        (total,lat) => {return total+lat;}, 0.0
      )/tracksPositionNotNull.length;

    const lng = tracksPositionNotNull
      .map(track=>{
        return parseFloat(track.position.longitudeDegrees);
      }).reduce(
        (total,lng)=> {return total+lng;},0.0
      )/tracksPositionNotNull.length;

    return { 
      lat,
      lng
    } 
  }