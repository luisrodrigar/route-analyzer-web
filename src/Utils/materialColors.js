import {red, pink, purple, deepPurple, indigo, blue,
        lightBlue, cyan, teal, green, lightGreen, 
        lime, yellow, amber, orange, deepOrange, brown, grey, blueGrey  } from 'material-ui/colors';

const mainColors = [
  'Red',
  'Pink',
  'Purple',
  'Deep Purple',
  'Indigo',
  'Blue',
  'Light Blue',
  'Cyan',
  'Teal',
  'Green',
  'Light Green',
  'Lime',
  'Yellow',
  'Amber',
  'Orange',
  'Deep Orange',
  'Brown', 
  'Grey',
  'Blue Grey'
];
const lightPalette = [200];
const darkPalette = [500,900];

function getColorObject(name){
  switch(name){
    case mainColors[0]:
    return red;
    case mainColors[1]:
    return pink;
    case mainColors[2]:
    return purple;
    case mainColors[3]:
    return deepPurple;
    case mainColors[4]:
    return indigo;
    case mainColors[5]:
    return blue;
    case mainColors[6]:
    return lightBlue;
    case mainColors[7]:
    return cyan;
    case mainColors[8]:
    return teal;
    case mainColors[9]:
    return green;
    case mainColors[10]:
    return lightGreen;
    case mainColors[11]:
    return lime;
    case mainColors[12]:
    return yellow;
    case mainColors[13]:
    return amber;
    case mainColors[14]:
    return orange;
    case mainColors[15]:
    return deepOrange;
    case mainColors[16]:
    return brown;
    case mainColors[17]:
    return grey;
    case mainColors[18]:
    return blueGrey;
    default:
    return getColorObject(mainColors[Math.floor(Math.random()*mainColors.length)]);
  }
}

function getColor(){
  const colorDark = darkPalette[Math.floor(Math.random()*darkPalette.length)];
  const colorLight = lightPalette[Math.floor(Math.random()*lightPalette.length)];
  const colorName = mainColors[Math.floor(Math.random()*mainColors.length)];

  const colorObject = getColorObject(colorName);

  let colorValue = colorObject[colorDark];
  let colorLightValue = colorObject[colorLight];

  return [colorValue,colorLightValue];
}

export function getColorRandom(numColors, currentColors){
  let arrayColors = [];
  let color = null;
  for(let i=0; i<numColors; i++){
    if(numColors<(mainColors.length * darkPalette.length)){
      do{
        color = getColor();
      }while(currentColors.map(x=>x.join('')).indexOf(color.join(''))>=0);
      currentColors.push(color);
    }else{
      color = getColor();
    }
    arrayColors.push(color);
  }
  return arrayColors;
}