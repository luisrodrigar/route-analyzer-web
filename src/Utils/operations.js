import {getColorRandom} from './colors';
import * as date        from 'datejs';

export function getLapsTrackPoints(laps){
  return laps.map(lap => {
      return {
            index:lap.index,
            tracks:getLapTrackPoint(lap),
            color:lap.color,
            lightColor: lap.lightColor,
            startTime: lap.startTime ?  parseDate(lap.startTime).getTime() : null,
            totalTime: lap.totalTimeSeconds,
            distance: lap.distanceMeters,
            maxSpeed: lap.maximumSpeed,
            avgSpeed: lap.averageSpeed,
            avgBpm: lap.averageHearRate,
            maxBpm: lap.maximumHeartRate,
            cal: lap.calories,
            intensity: lap.intensity
      }
  });
}

function getLapTrackPoint(lap){
  return getLapTracksWithPosition(lap).map(trackPoint => {
      const lat = parseFloat(trackPoint.position.latitudeDegrees);
      const lng = parseFloat(trackPoint.position.longitudeDegrees);
      const position = {lat, lng};
      const alt = parseFloat(trackPoint.altitudeMeters);
      const speed = parseFloat(trackPoint.speed);
      const date = trackPoint.date ? parseDate(trackPoint.date).getTime() : null;
      const dist = parseFloat(trackPoint.distanceMeters);
      const bpm = parseInt(trackPoint.heartRateBpm,10);
      const index = parseInt(trackPoint.index,10);
      return {
        index,
        date,
        position,
        alt,
        speed,
        dist,
        bpm
      };
  });
}

function parseDate(dateString){
  return Date.parse(dateString);
}

export function setLapColors(laps){
  let currentColors = laps
      .filter(lap => lap.color && lap.lightColor)
      .map(lap => [lap.color,lap.lightColor]);
  let numLapsWithoutColor = laps
      .filter(lap=> !lap.color && !lap.lightColor)
      .length;
  const colors = getColorRandom(numLapsWithoutColor, currentColors);
  laps.filter(lap => !lap.color && !lap.lightColor)
      .forEach((lap, index) => {
        lap.color = colors[index][0];
        lap.lightColor = colors[index][1];
      });
}

export function getLapsSpeed(laps){
  return laps.map(lap=> {
    return {
      index: lap.index,
      color: lap.lightColor,
      label: "Lap " + lap.index,
      stroke: 2,
      tracks:lap.tracks.map(track=>[track.date, track.speed])
    }
  });
}

export function getLapsHeartRate(laps){
  return laps.map(lap=> {
    return {
      index: lap.index,
      color: lap.lightColor,
      label: "Lap " + lap.index,
      tracks: lap.tracks.map(track => [track.date, track.bpm])
    }
  });
}

export function getLapsElevations(laps){
  return laps.map(lap=> {
    return {
      index: lap.index,
      color: lap.color,
      label: "Lap " + lap.index,
      tracks: lap.tracks.map(track => [track.date, track.alt])
    }
  });
}

export function calculateAvgValues(laps, avgMethod) {
    return laps
        .map(lap => [getAvgDate(lap.tracks), avgMethod(lap)]);
}

function getAvgDate(tracks) {
    return (tracks[0].date + tracks[tracks.length-1].date) / 2;
}

function calculateAvgSingleLap(laps, avgArray, avgMethod) {
    let avgDate = avgArray[0][0];
    let initDate = laps[0].tracks[0].date;
    let endDate = laps[0].tracks[(laps[0].tracks.length)-1].date;
    avgArray = [];
    avgArray.push([initDate, avgMethod(laps[0])]);
    avgArray.push([avgDate, avgMethod(laps[0])]);
    avgArray.push([endDate, avgMethod(laps[0])]);
    return avgArray;
}

export function getAvgBpm(laps){
    let avgBpmMethod = function(lap) {
        return lap.avgBpm;
    };
    let avgBpm = calculateAvgValues(laps, avgBpmMethod);
    return laps.length > 1 ? avgBpm : calculateAvgSingleLap(laps, avgBpm, avgBpmMethod);
}

export function getAvgSpeed(laps){
    let avgSpeedMethod = function(lap) {
        return lap.avgSpeed;
    };
    let avgSpeed = calculateAvgValues(laps, avgSpeedMethod);
    return laps.length > 1 ? avgSpeed : calculateAvgSingleLap(laps, avgSpeed, avgSpeedMethod);
}

export function getHeartRateData(laps) {
  return laps.flatMap(lap =>
      lap.tracks.map(track => [track.date,track.bpm]));
}

export function getElevationData(laps) {
  return laps.flatMap(lap =>
    lap.tracks.map(track => [track.date,track.alt]));
}

export function getSpeedData(laps){
  return laps.flatMap(lap =>
      lap.tracks.map(track => [track.date,track.speed]));
}

function getLapTracksWithPosition(lap){
  return lap.tracks
    .filter(track => {
      if(track && track.position)
        return track;
      else return null;
    });
}

export function getNearestPosition(laps,position){
  let indexLap =null, indexTrackpoint = null, minDistance = Number.POSITIVE_INFINITY;
  laps.forEach((lap, eachIndexLap)=>{
    lap.tracks.forEach((track,eachIndexTrackpoint)=>{
      let eachPosition = {
        lat: track.position.lat,
        lng: track.position.lng,
      };
      let currentMin = getDistanceBetweenPoints(eachPosition,position);
      if(currentMin<minDistance){
        indexTrackpoint = eachIndexTrackpoint;
        indexLap = eachIndexLap;
        minDistance = currentMin;
      }
    })
  })
  return {
    indexLap,
    indexTrackpoint
  }
}

function getDistanceBetweenPoints(point1, point2){
  // Convert degrees to radians
  let latP1 = degrees2Radians(point1.lat), lngP1 = degrees2Radians(point1.lng);
  let latP2 = degrees2Radians(point2.lat), lngP2 = degrees2Radians(point2.lng);

  // Radius of earth in meters
  let earthRadiusMeters = 6378100;

  // Point P 
  let rho1 = earthRadiusMeters * Math.cos(latP1);
  let z1 = earthRadiusMeters * Math.sin(latP1);
  let x1 = rho1 * Math.cos(lngP1);
  let y1 = rho1 * Math.sin(lngP1);

  // Point Q
  let rho2 = earthRadiusMeters * Math.cos(latP2);
  let z2 = earthRadiusMeters * Math.sin(latP2);
  let x2 = rho2 * Math.cos(lngP2);
  let y2 = rho2 * Math.sin(lngP2);

  // Dot product
  let dot = (x1 * x2 + y1 * y2 + z1 * z2);
  let cosTheta = dot / (Math.pow(earthRadiusMeters, 2));

  let theta = Math.acos(cosTheta);

  return earthRadiusMeters * theta;

}

function degrees2Radians(degree){
  return degree * Math.PI / 180.0;
}
