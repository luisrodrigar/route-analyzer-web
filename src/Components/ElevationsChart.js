import React    from 'react';
import sizeMe from 'react-sizeme';
import Grid from 'material-ui/Grid';
import AreaChart from './Charts/AreaChart'
import {getElevationData, getLapsElevations} from '../Utils/lapOperations';

// Using a react-sizeme library to get width and height values => Handling on resize event
const MySizeGrid = sizeMe({ monitorHeight: true })(Grid);

export class ElevationsChart extends React.Component {

  constructor(props){
    super(props);
    this.handleChangeSizeContent = this.handleChangeSizeContent.bind(this);
    let elevationsData = getElevationData(this.props.laps);
    let lapsElevations = getLapsElevations(this.props.laps);
    this.state = {
      elevations: elevationsData,
      laps:   lapsElevations,
      width:  500,
      height: 200,
    };
  }

  componentDidUpdate(prevProps, prevState){
    if(prevProps.laps !== this.props.laps){
      let chart = this.chartGrid;
      let elevationsData = getElevationData(this.props.laps);
      let lapsElevations = getLapsElevations(this.props.laps);
      this.setState({
        elevations: elevationsData,
        laps: lapsElevations,
        width:chart.state.width,
        height:chart.state.height,
      });
    }
  }

  handleChangeSizeContent(){
    let chart = this.chartGrid;
    this.setState({
      width:chart.state.width,
      height:chart.state.height,
    });
  }

  componentDidMount(){
    window.addEventListener("resize", this.handleChangeSizeContent);
  }

  componentWillUnmount(){
    window.removeEventListener("resize", this.handleChangeSizeContent);
  }

	render(){
		return(
      <MySizeGrid 
            ref={ref=>this.chartGrid = ref}
            item
            xs={4}
            style={{maxHeight:'500px', minWidth:'420px'}}>
            <AreaChart  data={this.state.elevations} 
                        laps={this.state.laps}
                        textYAxis={this.props.yTitle}
                        textXAxis={this.props.xTitle}
                        padding = {40}
                        width = {this.state.width}
                        height = {this.state.height}/>
      </MySizeGrid>
		);
  }

}

export default ElevationsChart;