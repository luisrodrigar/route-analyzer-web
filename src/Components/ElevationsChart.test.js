import React from "react";
import { ElevationsChart} from "./ElevationsChart";
import sizeMe from 'react-sizeme';
import { mount, shallow } from "enzyme";
jest.mock('../Utils/operations');
jest.mock('../actions/index');
import * as operations from '../Utils/operations';
import Grid from 'material-ui/Grid';
import toJson from "enzyme-to-json";
import AreaChart from "./Charts/AreaChart";


describe("Elevations Chart Test", () => {
    let elevations;
    let elevationsLap;
    let updateTrackPointMock;
    let currentTrackPoint;
    let laps;
    let component;
    let props;
    beforeAll(() => {
        currentTrackPoint = {
            index: 0,
            position: {
                lat: 1.2123,
                lng: 0.2324
            },
            alt: 12.343
        };
        laps = [
            {
                index: 0,
                tracks: [
                    {
                        index: 0,
                        position: {
                            lat: 1.2123,
                            lng: 0.2324
                        },
                        alt: 12.343
                    },
                    {
                        index: 1,
                        position: {
                            lat: 1.2122,
                            lng: 0.2325
                        },
                        alt: 12.543
                    }
                ]
            },
            {
                index: 1,
                tracks: [
                    {
                        index: 0,
                        position: {
                            lat: 1.2124,
                            lng: 0.2327
                        },
                        alt: 13.343
                    },
                    {
                        index: 1,
                        position: {
                            lat: 1.2128,
                            lng: 0.2326
                        },
                        alt: 15.543
                    }
                ]
            }
        ];
        elevations = [
            ["2019-07-07T10:34:00Z", 123],
            ["2019-07-07T10:34:10Z", 130],
            ["2019-07-07T10:34:20Z", 132],
            ["2019-07-07T10:34:30Z", 114],
            ["2019-07-07T10:34:40Z", 124]
        ];
        elevationsLap = [
            {
                index: 0,
                color: '#032',
                label: "Lap0",
                tracks: [
                    ["2019-07-07T10:34:00Z", 123],
                    ["2019-07-07T10:34:10Z", 130],
                ]
            },
            {
                index: 1,
                color: '#012',
                label: "Lap1",
                tracks: [
                    ["2019-07-07T10:34:20Z", 132],
                    ["2019-07-07T10:34:30Z", 114],
                    ["2019-07-07T10:34:40Z", 124]
                ]
            },
            {
                index: 2,
                color: '#002',
                label: "Lap2",
                tracks: [
                    ["2019-07-07T10:34:40Z", 124]
                ]
            }
        ];
        props = {
            currentTrackPoint: currentTrackPoint,
            laps: laps,
        };
    });
    beforeEach(() => {
        operations.getElevationData.mockImplementationOnce(() => elevations);
        operations.getLapsElevations.mockImplementationOnce(() => elevationsLap);
        updateTrackPointMock = jest.fn();
    });
    afterEach(() => {
        jest.resetAllMocks();
        component.unmount();
    });
    it("Checking the snapshot", async () => {
        component = shallow(<ElevationsChart/>);
        expect(toJson(component)).toMatchSnapshot();

    });
    it("Checking the resize event on window.", async () => {
        const mockSizeGrid = sizeMe({monitorHeight: true})(Grid);
        const widthMockValue = 15;
        const heightMockValue = 10;
        mockSizeGrid.state = {
            width: widthMockValue,
            height: heightMockValue
        };
        component = mount(<ElevationsChart props={props}
                                           yTitle={'Altitude (m)'}
                                           xTitle={'Time (hh:mm:ss)'}
                                           updateTrackpoint={updateTrackPointMock}/>);
        const instance = component.instance();
        instance.chartGrid = mockSizeGrid;
        window.dispatchEvent(new Event('resize'));
        expect(instance.state.width).toEqual(widthMockValue);
        expect(instance.state.height).toEqual(heightMockValue);
    });
    it("Checking the resize event on window.", async () => {
        component = shallow(<ElevationsChart props={props}
                                             yTitle={'Altitude (m)'}
                                             xTitle={'Time (hh:mm:ss)'}
                                             updateTrackpoint={updateTrackPointMock}/>);
        const areaChart = component.childAt(0).find(AreaChart).first();

        areaChart.simulate('mouseOver');

        expect(updateTrackPointMock).toBeCalled();
    });
});