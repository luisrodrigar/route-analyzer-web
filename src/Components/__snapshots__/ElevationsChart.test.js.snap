// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elevations Chart Test Checking the snapshot 1`] = `
<SizeMe(WithStyles(GridWrapper))
  item={true}
  style={
    Object {
      "maxHeight": "200px",
      "minWidth": "500px",
    }
  }
  xs={12}
>
  <AreaChart
    data={
      Array [
        Array [
          "2019-07-07T10:34:00Z",
          123,
        ],
        Array [
          "2019-07-07T10:34:10Z",
          130,
        ],
        Array [
          "2019-07-07T10:34:20Z",
          132,
        ],
        Array [
          "2019-07-07T10:34:30Z",
          114,
        ],
        Array [
          "2019-07-07T10:34:40Z",
          124,
        ],
      ]
    }
    height={200}
    laps={
      Array [
        Object {
          "color": "#032",
          "index": 0,
          "label": "Lap0",
          "tracks": Array [
            Array [
              "2019-07-07T10:34:00Z",
              123,
            ],
            Array [
              "2019-07-07T10:34:10Z",
              130,
            ],
          ],
        },
        Object {
          "color": "#012",
          "index": 1,
          "label": "Lap1",
          "tracks": Array [
            Array [
              "2019-07-07T10:34:20Z",
              132,
            ],
            Array [
              "2019-07-07T10:34:30Z",
              114,
            ],
            Array [
              "2019-07-07T10:34:40Z",
              124,
            ],
          ],
        },
        Object {
          "color": "#002",
          "index": 2,
          "label": "Lap2",
          "tracks": Array [
            Array [
              "2019-07-07T10:34:40Z",
              124,
            ],
          ],
        },
      ]
    }
    legend="m"
    padding={40}
    ticks={5}
    width={1400}
  />
</SizeMe(WithStyles(GridWrapper))>
`;
